### ISSUE TITLE
Provide a way to rotate Logstash's internal logs

### ISSUE THREAD

**Issue by mhaji** [elastic/logstash#5016]
I'm using the `--log` to redirect Logstash's logs to a file rather than `STDOUT`. I've noticed that if there are errors that this file can grow uncontrollably. I've not found an option to be able to rotate it or control the size of this log file.

Could we do something similar to the output `file` plugin whereby we can specify a date-based log rotation via the joda time format, for example, `--log "./test-%{+YYYY-MM-dd}.txt"`?

**Comment by purbon** [comment]
might be you can configure logrotate http://www.linuxcommand.org/man_pages/logrotate8.html ?

**Comment by jordansissel** [comment]
Logstash used to respond to SIGHUP to reopen it's logfile which would work when combined with logrotate, but I think that feature (reopen log) was accidentally removed when config reloading was added.

**Comment by BellBiker** [comment]
I am using the ELK stack on a windows system and I also recognized  a problem with the missing possibility to control the size of the logstash internal log file. In a case where the harvested log files are producing a huge amount of log file rows within short time and therefore logstash documents a slowdown of the pipeline this causes also an uncontrollable increasing logstash logfile. At least logstash itself crashes due to the fact that the drive, where the log file is located, gets full. 

Therefore I would prefer if logstash itself, maybe conigurable, cares about a log rotation of his own log files. The above proposal will not help in cases where the log file increases within minutes or latest hours as it happens in my case.

**Comment by fgimian** [comment]
@jordansissel Hey there, we are also facing this issue in our production environment.

Unfortunately, logrotate is problematic because our Logstash logs get very large very fast (around 10 GB per day) and when attempting hourly rotation of the logs, it appears that Logstash stops ingesting logs until this completes.  Most of the log entries are indeed warnings which we are working through, but it would be great to have a way around this in the meantime.

Do you think we will see the return of SIGHUP soon?  This would resolve the issue for us as we can simply write a cron to send the signal to rotate the log on an hourly basis.

Thanks so much!
Fotis

**Comment by jsvd** [comment]
@fgimian hi, today I'll work a patch to reintroduce reopening the log files on SIGHUP

on a side note, could you provide some information on the kind of log lines being produced by logstash? are they errors? maybe superfluous debug information being logged as info/warn?

**Comment by fgimian** [comment]
Thanks so much @jsvd, yeah the lines we are seeing are primarily grok failures which we are working on resolving with our Elasticsearch consultant Nic.  These are indeed errors that we have reduced, but sadly the app producing our logs is a bit of a pain and often produces lines that we need to write new grok patterns for.  So we would love to have this feature return at least as a precaution 😄

**Comment by jsvd** [comment]
That's strange, looking at the code I don't see warnings being printed when grok fails to parse. Are you running logstash with --debug in production?
Can you provide an example or 2? thank you

**Comment by fgimian** [comment]
The messages are hugely long and do contain quite a lot of company specific info so I've shortened it to show you the most important parts:

```
{:timestamp=>"2016-06-29T18:18:58.143000+1000", :message=>"Failed action. ", :status=>400, :action=>["index", {:_id=>nil, :_index=>"mailapp-imapserv-2016.06.29", :_type=>"imapserv", :_routing=>nil}, #<LogStash::Event:0x63db44e8 @metadata_accessors=#<LogStash::Util::Accessors:0x5a14388d @store={}, @lut={}>, @cancelled=false, @data=......, :response=>{"create"=>{"_index"=>"mailapp-imapserv-2016.06.29", "_type"=>"imapserv", "_id"=>"AVWbPV1w47cScVTxxAcf", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"Field name [CASSOBJPOOL\\3a\\3agetSliceWithColmNames\\3a  CassHost\\3ax.x.x.x, CassPort\\3a9163 expected message uuid] cannot contain '.'"}}}, :level=>:warn}
```

Happy to raise a case so I can provide further info, but our consultant from Elasticsearch did mention that this was due to entries in logs that simply were not accounted for in the grok patterns, and he has helped us significantly improve our Logstash rules such that this problem may no longer affect us enough to warrant regular log rolling.

Either way, would be awesome to have this option if we ever bump into a new log type and encounter similar problems in the future 😄 

Hope that info helps
Fotis

**Comment by suyograo** [comment]
Will be addressed in #5650
